### ISSUE TITLE
more flexibility for chooser options

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#52]
**Reported by @cristianbogdan on 24 Oct 2002 23:32 UTC**
besides having supplementary choices (see bug 59), sometimes there is a need to
limit the number of choices, to combine sets of choices, etc. currently such
things need be generated manually in html and (as in bug 59) need java scriplets

probably the most flexible syntax to solve this is:

<mak:input name=... dataType="set a.b.C" >
      <mak:option value="please choose">
      <mak:list from= anything where= anything/>
         <mak:option expr=somethingOfType(a.b.C) />
      </mak:list>
      <mak:option value="-----------">
      <mak:option value="---outside Europe---">
      <mak:list from= anythingElse where= anythingElse/>
         <mak:option expr=somethingOfType(a.b.C) />
      </mak:list>
</mak:input>

Migrated-From: http://trac.makumba.org/ticket/60

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 25 Oct 2002 00:19 UTC**
there seems to be another problem here: the actual option shown, which is not
always the "title field" of the MDD (which should be replaced by an OQL function
anyway)

also, i misplaced some closing slashes in the example above, so here it comes again:

<mak:input name=... dataType="set a.b.C" >
      <mak:option value="please choose"/>
      <mak:list from= anything where= anything>
         <mak:option expr=somePtrToType(a.b.C)>
            <mak:value expr=some title>
         </mak:option>
      </mak:list>
      <mak:option value="-----------">
      <mak:option value="---outside Europe---">
      <mak:list from= anythingElse where= anythingElse>
         
         <mak:option expr=somePtrToType(a.b.C) />
      </mak:list>
</mak:input>

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 9 Jun 2003 12:05 UTC**
moving text of bug 59 here; also marking it dependent on page analysis, it was
impossible to fix this when training loops were on

currently the makumba choosers contain strictly the members of the respective
value set. this is problematic: if one wants to add a dummy choice (meaning
null, or indicating something to the user), one has to build the chooser "by
hand" in html

the proposals below suggest to add the null option in the MDD. that works for
int{} but doesn't work for set or pointer chooser. i for one would suggest not
to couple the MDD with the view so closely. i would have this just as a
<mak:input > feature somehow. like e.g.
<mak:input name=fieldname topChoices="please choose:, undecided"/>

WORKAROUNDS:
generating HTML by hand is possible, for all of int{}, set, ptr. that involves
writing <mak:input name=fieldname type=hidden /> and writing the options by hand
(for int{}) or generating them by mak:list (for set and ptr)

ORIGINAL COMMENTS
Fred:
I think the standard interface element for enumerated int fields (like
"0=no,1=yes") in makumba, which present as a dropdown box (as translation
for mak:editform) should allow to set the value to "NULL", if null is an
allowed value for the data.

Or there could be parameter on the mak:editform that controls whether
"blank" should be a listed value in the dropdown box. 


I think this makes sense, because if you look at the MDDs, the value
"null" often has an important functional meaning in the application. (one
could philosophise whether that is software-engineering-desirable, or just
overly techie-saving-space?)

Furthermore, the dropdown edit box now already wants to "default" to the
actual value of the field, but this is not possible when the value is
NULL, but Null is not included on the dropdown list. 

I know we have to be careful HOW to present (phrase) the "null" option, as
the actual phrase should make sense to the end-user who uses the
application.

Stefan:
The same problem is occuring with pointers (eg Afgan citizenship phenomenon)

how about giving the meaning of the null value directly in the MDD, since
that is the place were all those comments are ending up anyway:
gender=int{"undecided"=null, "male"=0, "female"=1} ; Gender
tshirtsize=int{"average"=null, "S"=0, "M"=1,"L"=2,"XL"=3}
but this means that the field can be nullable - is it ok?

or should there be a "not null" (or so) field type modifier that would
specify that the null option would only be shown (and preselected) if the
current value is null.
tshirtsize=not null int{"average"=null, "S"=0, "M"=1,"L"=2,"XL"=3}
resulting in: once you have chosen your tshirt size, you can't set it to
"average" anymore

the first option would be the default one, as it is now.

It is a bigger change for pointers. now we have
citizenship=ptr general.Country;    Citizenship
any ideas where to specify the null value?
Maybe just compute it from "[ Choose "+description+" ]" that woud result in
"[ Choose Citizenship ]" for the example above. That principle could also be
used for int and char enum fields instead of the one proposed above, which
eliminates nullable/not nullable problem (everything remains not nullable),
but null values get generic descriptions (eg: "[ Choose Gender ]") .

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 9 Jun 2003 12:06 UTC**
*** Bug 59 has been marked as a duplicate of this bug. ***

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 30 Jun 2003 12:04 UTC**
besides having null options (bug 59, which also goes for other types than ptr
and set, which is why i re-separated it), is there anything more to add to the
design at comment 1?

my goal is to have something powerful enough to be able to produce something
like the "subgroups" chooser at
https://parade.best.eu.org/mak-test-k/groups/admin/groupEdit.jsp?team=aj83cuo

(there you have the selected stuff at the top, which could be achieved e.g. by
displaying options for the current set first, then displaying options for all
the stuff that is not in the set afterwards, i.e. "whether the option is
selected or not" should be accesible to the programmer)

i have a question: was there ever a need for null members in sets? i think
<option value="nil" /> should not be supported for sets

**Comment by ghost** [comment]
**Comment by @stefanb on 20 Oct 2003 18:13 UTC**
>i have a question: was there ever a need for null members in sets? i think
><option value="nil" /> should not be supported for sets
I don't think there ever was such need, and it doesn't seem natural either.

I am still skeptical about those
      <mak:option value="please choose"/>
Which would still have to be sth like
      <mak:option value="nil">please choose</mak:option>
which saves very little programming efforts comparing to standard html tag:
      <option value="">please choose</option>

Your example:
<mak:input name=... dataType="set a.b.C" >
      <mak:option value="please choose"/>
      <mak:list from= anything where= anything>
         <mak:option expr=somePtrToType(a.b.C)>
            <mak:value expr=some title>
         </mak:option>
      </mak:list>
      <mak:option value="-----------">
      <mak:option value="---outside Europe---">
      <mak:list from= anythingElse where= anythingElse>
         
         <mak:option expr=somePtrToType(a.b.C) />
      </mak:list>
</mak:input>

Can be with no extra complexity written with existing features as:
<mak:input name=... dataType="set a.b.C" display="hidden"/>
<input name=... >
      <option value="">please choose</option>
      <mak:list from= anything where= anything>
         <option value="<mak:value expr=somePtrToType(a.b.C)">
            <mak:value expr=some title>
         </option>
      </mak:list>
      <option value="">-----------</option>
      <option value="">---outside Europe---</option>
      <mak:list from= anythingElse where= anythingElse>
         
         <option value="<mak:value expr=somePtrToType(a.b.C)" >
             <mak:value expr="somePtrToType(a.b.C).titleField" />
         </option>
      </mak:list>
</input>

or did i misunderstood sth important?

**Comment by ghost** [comment]
**Comment by @stefanb on 20 Oct 2003 18:51 UTC**
>or did i misunderstood sth important?
yes, i realizeed i did miss sth important after reading attachment 17:
in edit forms some optins are preselected (with current value):
  <option SELECTED>bla</option>
...but would mak:option be able to determine if it should be selected or not?
will this be the only benefit?

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 30 Dec 2004 14:48 UTC**
after having implemented quite many (as i call them) manual html drop-down
choosers, i have the following comments to make the chooser values more
customisable.

1.) do not make an extra tag, this just confuses, but use mak:input. 
some other mak:inputs have some attributes that apply only for special data
types, so i do not see any new problem by keeping mak:input, extended by some
attribute options.


2.) i see three different things that might be customised
a.) limit the values shown in the chooser.
b.) add some more options.
c.) a special sub-case of b., to allow null values


i have seen and used a lot cases a.) and c.) so far, and from my experience in
that, these are my proposals.

a.) this is pretty much like a where-clause in a mak:list, so i would propose to
add an attribute called something like "limitOptionsTo" or so.

c.) an attribute called something like allowNull="true/false"

b.) here, i am not yet sure, basically because so far, i did not come across any
example where i would need to add options besides the null, as in case c.).
possibilities could though be to allow a list of comma-seperated values, as well
as more complex things like an OQL query with an attribute
additionalOptionsFrom="OQL stuf...", maybe also with a where part.


it is a rather simple design, imho, from the user point of view, but it should
allow all for many customisations. 
i am not sure however how difficult it is to implement, and how difficult it is
to check where these attributes are allowed or not.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 30 Dec 2004 23:57 UTC**
i'm very glad to see you active here Rudi

about how to implement: 
- choser option computation is pretty well isolated in
org.makumba.controller.html, i think choiceEditor is the base class. the only
classes that generate queries now are ptrEditor and its subclass setEditor. 
- don't forget that the options have to be a set (i.e. an element appears at
most once, or at least it's submitted at most once)
- you also must ensure that the pointer chosers have at most one choice SELECTED
- and that set chosers have no nil choiches.

i encourage you to express your designs as imaginary makumba tags that solve a
certain problem. that will make you check if your design near completion, and
will make others understand.

then 
- i think there is a case (d) or a better expression of (b) which is "generate
options through more than 1 query". 
- if you indicate nil choice, is it obvious that it's going to be the first?
can't there be more of them?
- how about separators in the choser?
- you discuss "more flexibility" for _options_ but not for _option_titles_. you
could do that by saying title="this.name this.surname" where 'this' is the
default label for the choser query, and the title string is some combination of
makumba expressions and HTML... (the !title in the MDD, referred to e.g. in
attachment 17 is an unfortunate intrusion of view in the model so we skip it)

for an example of all these, i still think that the best problem for this bug is
the one at comment 5

> my goal is to have something powerful enough to be able to produce something
> like the "subgroups" chooser at
> https://parade.best.eu.org/mak-test-k/groups/admin/groupEdit.jsp?team=aj83cuo

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 31 Dec 2004 00:12 UTC**
so we need a flexible way to specify options, option titles, nil options and nil
option titles.

i start to realize that the options could be expressed by one or more mak:lists
that define a special label 'this' (or maybe better 'choice') which is the
pointer to the option. then all the text inside the mak:list is the option title.

<mak:input name="aChoiceField">
  <mak:list from="someType this, ..." ... >
     <mak:value expr="this.aField" /> bla <mak:value expr="this.aNotherFld" />
  </mak:list>
.... 
</mak:input>

one could argue that you shouldn't need to repeat the type of aChoiceField in
someType. at a later stage you could make a convention that the default FROM of
a mak:list inside a mak:input is "theTypeOfTheChoiceField this". 

some questions still remain: 
- how to express nil. one way is to simply declare that all non-whitespace text
that appears inside the <mak:input> and is not in a <mak:list> is the title of a
nil choice...
- how to force selection of some (nil or not nil) options. a simple solution is
to invent <mak:selected/>, placed anywhere within the option title.
- if you have more than one <mak:list> inside the <mak:input>, both will have to
define the this label. this may fux up the page analyzer, i'm not sure. but it's
easy to accept this1, this2...

an unexpected benefit of this design is that if the mak:input is somewhere
inside another mak:list or mak:object, it can cooperate with it (i.e. limit its
output based on what's selected there). 

ok, now going from user to developer :)... 

implementing this design is not that difficult, as it's based on the already
well-implemented mak:list. the strategy would be:
- mak:input needs be declared that it has a body in makumba.tld (thus starting
to fix bug 502). to satisfy that, org.makumba.view.jsptaglib.InputTag should
implement javax.servlet.jsp.tagext.BodyTag. that requires implementing
doInitBody() (which can be empty) and setBodyContent() (see below)
- the mak:list knows (from page analysis, in
org.makumba.view.jsptaglib.QueryTag.doStartAnalyze()) that it has a mak:input
ancestor (findAncestorWithClass(InputTag.class) ) and sees to it that two
markers are generated: <TITLE> before the first mak:value and </TITLE> after the
last mak:value inside the mak:list. these mak:values can simply decide at page
analysis to generate something supplementary in their output, and cache this
fact in their ValueComputer
- the first mak:value in the mak:list also supplements itself to print the
effect of <OPTION> <mak:value expr="this" /></OPTION> before it prints <TITLE>
(this needs be done at page analysis when queries are cached)
- the mak:input gets hold of its body content (by defining setBodyContent()) and
parses the body content for the OPTION and TITLE markers. the InputTag is
already prepared for that, see line 169. 
- in the body content, everything outside the markers is considered nil separators
- whenever a title or a nil separator contains <mak:selected/> (whose
implementation can simply generate a <SELECTED/> marker) that title's option is
declared selected...
- once it has its choices, the mak:input passes them normally to
org.makumba.controller.html.choiceEditor. currently this is done in
org.makumba.view.jsptaglib.InputTag, line 189:

    String formatted=getForm().responder.format(name, type, val, params,
extraFormatting.toString());

probably a new parameter "choiches" needs to be introduced for
org.makumba.controller.html.FormResponder.format(), a vector of triplets
(option, title, selected) that would only be used by
org.makumba.controller.html.choiceEditor.

i'm actually pretty excited about this design, but as you can guess i made the
long explanation above because i'd like to see you Rudi implementing it :)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 31 Dec 2004 00:39 UTC**
more about comment 10:
- of course you have to see to it that org.makumba.controller.html.ptrEditor
does not execute the default query at line 36
- of course mak:object will work just as well as mak:list inside a mak:input.
may be useful for expressing strange isolated choiches.
- you can skip whole iterations of the mak:list by using <c:if> and skipping the
first and the last mak:value
- due to the value of the pointer or set, some options will appear selected. for
simetry you might want <mak:selected selected="false" /> to block that
- for sets <mak:input> already generates a query to find out the selected elements.
- if you want a set's choices to show "selected elements first" (like in the
"subgroups" choser i mentioned at comment 5) you need to do

<mak:object from="someType a">
<mak:editForm object="a">
   <mak:input name="aSetField">
      <mak:list from="a.setField this">....</mak:list> 
      <mak:list from="theSetType this">....</mak:list>
   </mak:input>
</mak:editForm>
</mak:object>

elements already displayed by the first mak:list will not be shown again by the
second mak:list. this would be implemented by InputTag or ptrEditor eliminating
option duplicates

i guess if duplicates wouldn't be eliminated, you could always have (without
using subqueries...) 

      <mak:list from="a.setField this">....</mak:list> 
      <mak:list from="theSetType this, a.setField t2" 
                where="this<> t2" distinct="true">....</mak:list>

but distinct is not implemented yet (bug 713), plus i guess the result of the
join "theSetType this, a.setField t2" is pretty large

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 31 Dec 2004 02:16 UTC**
comment 10:
> - the mak:list knows (from page analysis, in
> org.makumba.view.jsptaglib.QueryTag.doStartAnalyze()) that it has a mak:input
> ancestor (findAncestorWithClass(InputTag.class) ) and sees to it that two
> markers are generated: <TITLE> before the first mak:value and </TITLE> after 
....
> - the first mak:value in the mak:list also supplements itself to print the
> effect of <OPTION> <mak:value expr="this" /></OPTION> before it prints <TITLE>
> (this needs be done at page analysis when queries are cached)


actually markers are easier to implement in the mak:list (QueryTag) itself
rather than in the mak:values inside it. at analysis time it can cache the
"this" expression in the query (maybe making a ValueComputer for it) and at
runtime it can detect that it's inside a <mak:input> and print the "this"
expression and the markers before and after each iteration

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 Jan 2005 00:42 UTC**
fixed in 0.5.13
see the "Selector option display" section of the latest makumba doc

<mak:input dataType="T"...>
    this is a null option 
    <mak:list from="T choice"... orderBy...>this is a non-null option <mak:value
expr="choice.someField"/> 
    </mak:list>
    this is another null option, e.g. a separator
    <mak:list from="T choice"...>this is a non-null option <mak:value .../> 
    </mak:list>
</mak:input>

**Comment by ghost** [comment]
**Comment by @gwenael-alizon on 2 Jan 2005 17:29 UTC**
I find the current design very elegant, great work Cristi.
One restriction of its is however that it can only be used for inputing pointer
type fields as the choices need currently to be defined by a label ("choice").

I'm available if examples are needed to clarify the need.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 Jan 2005 18:41 UTC**
by "pointer type" you mean ptr and set i guess.

i guess you'd like to use it for int{....} fields too. actually could be used
for simple int or char fields. 

but those you don't necessarily need to express via mak:list, it may be that you
want to do a <c:foreach > or so... the advantage of the mak:list is that, being
a makumba tag i.e. under our control, it can announce the <mak:input> "hey, i
have a new iteration here, that probably means a new choice". maybe a new tag
<mak:optionSeparator/> can do the same job.

i guess we can add to the current design (please feedback):
- a choiceExpression for mak:list (default now is "choice" but for e.g. ints,
you will want label.intField)
- ability to separate options without using mak:list (mak:optionSeparator)
- ability to specify the option value without using a mak:list (proposal below:
set the choice attribute using c:set )

<mak:input ...>
nullOption
<mak:optionSeparator/>
<c:set var="choice" expr="whatever" /> nonNullOption 
<mak:optionSeparator/>

<c:foreach ...>
<c:set var="choice" expr="whatever" /> iteratedOption 
<mak:optionSeparator/>
</c:foreach>

separatorNull
<mak:list choiceExpression="a.intField" from="Type a"... >

<%-- mak:optionSeparator is default before the mak:list and after each iteration
--%>
</mak:list>
nullOption
</mak:input>

**Comment by ghost** [comment]
**Comment by @gwenael-alizon on 2 Jan 2005 20:25 UTC**
Commenting Cristi's proposal for addition to current design (comment 15)
-- - -- -- - -- -- - --
[guess we can add to the current design (please feedback):
- a choiceExpression for mak:list (default now is "choice" but for e.g. ints,
you will want label.intField) [A](Cristi]
i)
- ability to separate options without using mak:list (mak:optionSeparator) [ability to specify the option value without using a mak:list (proposal below:
set the choice attribute using c:set ) [C](B]
-)
-- - -- -- - -- -- - --

[Exactly what I would need
http://parade.best.eu.org/gwen-k/general/survey/user/editAnswer.jspx#105
   <mak:input field="intAnswer"> (int field)
      -- No answer --
      <mak:list from="general.survey.QuestionType questionType, 
                      questionType.choices choices"  
                     (choices.position choice not possible to set) 
              where="questionType=questionAnswered.type" 
              orderBy="choices.position" id="singleSelect"> 
         <mak:value expr="choices.content"/>
      </mak:list> 
   </mak:input>

[B](A]) Not sure of the use of having this extra tag:
We can already now put whatever we want as separator, and if it chosen it est
the value to null. Not so strange behavior. A benefit of having it would be that
user doesn't reset data to null by mistake when choosing a separator

[Interesting to have I guess. However I think it would be already interesting
for empty-body mak:inputs to accept EL for their value attribute


Finally my vote would go strangely enough to Cristi's first comment on this bug
:))) and using a mak:option

<mak:option value="" (dataType=""?)></mak:option>
Yes it's yet an extra tag but it:
-> makes not needed the possible </mak:optionSeparator> extra tag of point [B](C]),
as whatever you put outside the mak:option body is printed but not passed to the
mak:input
-> has HTML similarity 
-> accomodates point [C] especially if you can put java/EL in the mak:option 
-> of course let us have mak:lists outside the mak:options
-> accomodates defining label.x as the data to be inserted (x=intField or other)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 Jan 2005 21:19 UTC**
tnx for rapid feedback. this is how i like it. Rudi, are you around? others?
below follows a total turnaround of the situation :)

[is easy to implement in regard to choice generation (org.makumba.jsptaglib). 

however, it's pretty painful to oblige the org.makumba.controller.html level to
display a choser for an int... see below for a solution

btw, how urgent and often is this needed?

> empty-body mak:inputs to accept EL for their value attribute

you can already do it using
<c:set var=x ... />  then <mak:input ... value="$x" />
value is an OQL expression so it's hard for it to become EL.

and i don't see the connection of value with the option list. 

[B](A]) is needed already now. the only way to separate 2 null options is to put a
nonsensical mak:list/object. but mak:options solves it indeed.

> <mak:option value="" (dataType=""?)></mak:option>

dataType??? the data type of the option is pretty much decided by the mak:input.
i don't get what you want there.

well, yes, mak:option is another way to express it and is more HTML-familiar.
then the current mak:list will be equivalent to

<mak:list ...> <mak:option value="choice" >....</mak:option> </mak:list>

but the question is, what happens to the non-empty text outside any mak:option?
  is it an error? i think so (see below)

DON'T make too many pages with the non-mak:option design (current one) if you're
thinking to argue more for mak:option. pleae make a decission ASAP. it's pretty
hard to both maintain the present design (i.e. mak:option value=choice as
default) to coexist with mak:option. possible, but i somehow feel that if we
accept mak:option it should only be mak:option. not even for null choices which
would become

<mak:option> null choice </mak:option> 

also, do we agree that it's only option lists that will ever be inside <mak:input>?
or will you also want a HTML-familiar <mak:select> to indicate that?
<mak:input....> <%-- this means we're going to explicate the input --%>
    <mak:select> <%-- this means that the following is a set of options --%>
         <mak:option ...>... </mak:option>...
    </mak:select>
</mak:input>

again, you have to make the mak:select decission ASAP, before you have a lot of
code that needs change when you want to change design

having mak:select (or the promise that only option lists will be inside
mak:input, or at least that option lists are default), i can say "OK, this is
going to be a choser" and then force somehow the org.makumba.controller.html to
show a choser even for a normal int or char... but don't ask me how.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 2 Jan 2005 22:36 UTC**
some comments:

> [is easy to implement in regard to choice generation (org.makumba.jsptaglib). 
> 
> btw, how urgent and often is this needed?

i think the *only* place where i ever saw it is survey engine. though maybe
there it is a bad design to use the same field for more than one answer type (it
is used for a number question type as well as for the int-value of a choice,
where it should actually be the pointer to that choice).
so, i say, case [A](A]) should *not* influence our decision on the design, it is
rare (unless someone comes up with a second example)

>>empty-body mak:inputs to accept EL for their value attribute
> 
> you can already do it using
> <c:set var=x ... />  then <mak:input ... value="$x" />
> value is an OQL expression so it's hard for it to become EL.
> 
> and i don't see the connection of value with the option list. 


there is non.
 
> [is needed already now. the only way to separate 2 null options is to put a
> nonsensical mak:list/object. but mak:options solves it indeed.


i don't know if this came only from me asking if this is possible.. i just
tested a bit, and realised that this is not possible in the current design. the
question is - does anybody have any example where you need two null values after
each other?
 
> well, yes, mak:option is another way to express it and is more HTML-familiar.
> then the current mak:list will be equivalent to
> 
> <mak:list ...> <mak:option value="choice" >....</mak:option> </mak:list>


i'd see the main advantage in mak:option in the fact that it is more *readable*.
one clearly knows what will be an option.
 
> but the question is, what happens to the non-empty text outside any mak:option?
>   is it an error? i think so (see below)


i'd say it is an error, and either throw an exception saying that only tags can
be inside the mak:input, or just ignore non-tag text (i'd prefer option 1)
 
> DON'T make too many pages with the non-mak:option design (current one) if you're
> thinking to argue more for mak:option. pleae make a decission ASAP. it's pretty
> hard to both maintain the present design (i.e. mak:option value=choice as
> default) to coexist with mak:option. possible, but i somehow feel that if we
> accept mak:option it should only be mak:option. not even for null choices which
> would become
> 
> <mak:option> null choice </mak:option> 


i totally agree, and i did not do too much, only 3 pages, and this was just for
testing. and i could stop gwen from committing things :-)


> also, do we agree that it's only option lists that will ever be inside
<mak:input>?
> or will you also want a HTML-familiar <mak:select> to indicate that?
> <mak:input....> <%-- this means we're going to explicate the input --%>
>     <mak:select> <%-- this means that the following is a set of options --%>
>          <mak:option ...>... </mak:option>...
>     </mak:select>
> </mak:input>
> 
> again, you have to make the mak:select decission ASAP, before you have a lot of
> code that needs change when you want to change design


hm, this is a rather difficult decision, right now, i can't come up with any
clear idea. the only idea, not yet sure if it is related or not, would be that i
would like to have a null-option also for dates. [this however is a more
simplistic case than the chooser, i guess, cause it is either date or null to be
entered](B])

 
> having mak:select (or the promise that only option lists will be inside
> mak:input, or at least that option lists are default), i can say "OK, this is
> going to be a choser" and then force somehow the org.makumba.controller.html to
> show a choser even for a normal int or char... but don't ask me how.


i would not put a mak:select, but go for saying option lists is default.

so, to sum it up - pro <mak:option> is 
* the readability
* maybe(??) better suitable for a null-option for int{}

cons:
* if it is more difficult to implement.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 Jan 2005 23:53 UTC**
ok, do i hear that we go for mak:option and all non-blank text inside mak:input
that's outside mak:option is not legal?

this allows us to later implement char and int chosers, but we don't do it for
now but we can specify the intention in the doc. maybe also open a separate
enchancement bug (long bugs like this are hard to read)

also, not having mak:select means that if we ever want something else than
selectors inside mak:input, we'll need to invent a new tag for that purpose.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 3 Jan 2005 00:10 UTC**
> ok, do i hear that we go for mak:option and all non-blank text inside mak:input
> that's outside mak:option is not legal?


yes, thinking about it, it sounds good, makes it clearer in usage, and maybe in
implementing also.
 
> this allows us to later implement char and int chosers, but we don't do it for
> now but we can specify the intention in the doc. maybe also open a separate
> enchancement bug (long bugs like this are hard to read)


yes, i would go for an extra bug, this one is huge.
 
> also, not having mak:select means that if we ever want something else than
> selectors inside mak:input, we'll need to invent a new tag for that purpose.


but we would need to do that in any way if we have another purpose, no?

if you think it is safer for the future use, and if it is not too much
implementation overkill, then maybe a mak:select [maybe not called select?](but)
makes sense. really hard to decide, though.

and, would a null-option for a date affect this?

**Comment by ghost** [comment]
**Comment by @gwenael-alizon on 3 Jan 2005 00:54 UTC**
> ok, do i hear that we go for mak:option and all non-blank text inside 
> mak:input that's outside mak:option is not legal?

Go for me too

> also, not having mak:select means that if we ever want something else than
> selectors inside mak:input, we'll need to invent a new tag for that purpose

The ever sounds scary :p
.. but for now I can't think of other ways for inputing than selecting among
alternatives or directly putting a value in an empty-body mak:input.
And things not directly related to data input (layout?) can be either as
attribute of mak:input tag or outside it, or? Maybe I'm missing something? Tell me..
If we go though for this extra tag I would go for calling it mak:select for HTML
familiarity (unless Rudi has strong points/better proposals?).

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 Jan 2005 07:26 UTC**
> and, would a null-option for a date affect this?

what's a null option for a date?

about mak:select: i think it's easy to determine during analysis that since
there is mak:option inside the mak:input, we know we have a choser. so no need
for mak:select in the end i think.

**Comment by ghost** [comment]
**Comment by Lode.spruyt@BEST.eu.org on 3 Jan 2005 14:57 UTC**
From the (makumba) user point of view, just wanted to give you the feedback that
the mak:option is by far the most understandable idea raised in this topic. I'm
sure everyone in BEST/ITC is very interested in Rudi's/Gwen's survey engine
hacking, but untill now I doubted quit a lot whoever was gonna use it.
As for now I see something I would consider to use myself, so I think you're
going the right wat with the mak:option, for readability/usability.

just 1 cent

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 Jan 2005 16:48 UTC**
bug 784 deals with what's left here. design is finished for now anyway

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 4 Jan 2005 21:26 UTC**
>> and, would a null-option for a date affect this?
>
>what's a null option for a date?

i mean sometimes you have a date input field, but you do not want to enter a
date (that works now when you do not touch the input), or you mabye want to
reset a date to "not entered" --> that is not possible now, and that is what i
mean with a null option for a date.

maybe it would be enough to have a tick-box next to the input labelled "none" or
so, selected this the value would be reset to null.

any comments about this? and maybe we should continue this discussion (if
there's the need) in a new bug? bug 784 does not seem to fit exactly.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jan 2005 22:08 UTC**
> i mean sometimes you have a date input field, but you do not want to enter a
> date (that works now when you do not touch the input), or you mabye want to
> reset a date to "not entered" --> that is not possible now, and that is what i
> mean with a null option for a date.
> 
> maybe it would be enough to have a tick-box next to the input labelled "none" or
> so, selected this the value would be reset to null.

this is not even a choser issue, so it has little to do with bug 60. fill it
separately, i think Fred had other comments about dates, maybe look also there
